src/App.tsx

```diff
@@ -1,17 +1,148 @@
-import React from 'react';
-import './App.css';
-
-import users from './api/users';
+import React, { useState } from 'react';
+import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import { PreparedTodo } from './types/PreparedTodo';
+import { User } from './types/User';
+import { Todo } from './types/Todo';
+import { TodoList } from './components/TodoList/TodoList';
+import './App.scss';
 
 const App: React.FC = () => {
+  const users: User[] = usersFromServer;
+  const [todos, setTodos] = useState<Todo[]>(todosFromServer);
+  const defaultNewTodo = {
+    userId: 0,
+    title: '',
+    completed: false,
+  };
+
+  const [newTodo, setNewTodo] = useState({ ...defaultNewTodo });
+
+  const [isUserValid, setIsUserValid] = useState(true);
+  const [isTextValid, setIsTextValid] = useState(true);
+
+  const isNewTodoDataValid = () => {
+    let isDataValid = true;
+
+    if (newTodo.userId === 0) {
+      setIsUserValid(false);
+
+      isDataValid = false;
+    }
+
+    if (newTodo.title === '') {
+      setIsTextValid(false);
+
+      isDataValid = false;
+    }
+
+    return isDataValid;
+  };
+
+  const handleChange = (key: string, value: string | number) => {
+    setNewTodo(prevTodo => ({
+      ...prevTodo,
+      [key]: value,
+    }));
+  };
+
+  const handleAddClick = () => {
+    if (!isNewTodoDataValid()) {
+      return;
+    }
+
+    setTodos(prevTodos => [
+      ...prevTodos,
+      {
+        id: prevTodos[prevTodos.length - 1].id + 1,
+        ...newTodo,
+      },
+    ]);
+
+    setNewTodo({ ...defaultNewTodo });
+    setIsUserValid(true);
+    setIsTextValid(true);
+  };
+
+  const prepareTodos = (rawTodos: Todo[]): PreparedTodo[] => (
+    rawTodos.map(todo => ({
+      ...todo,
+      user: usersFromServer.find(user => user.id === todo.userId) || null,
+    }))
+  );
+
   return (
-    <div className="App">
-      <h1>Add todo form</h1>
+    <div className="app">
+      <h1 className="app__title">Add todo form</h1>
+
+      <form className="app__form">
+        <label className="app__input-group">
+          {!isTextValid && (
+            <span className="app__input-error">
+              <span>Please enter the title</span>
+              ⚠️
+            </span>
```

Comment:
Why do you need 2 spans here? fix it everywhere
```suggestion
            <span className="app__input-error">
              Please enter the title ⚠️
            </span>
```

Reply:
At some point I thought about the tests, but it doesn't work that way too, so I will remove it)